scripts/swap-beacon/src/adapter/odos.ts

```diff
@@ -0,0 +1,171 @@
+    const [priceIn, priceOut] = await Promise.all([
+      this.getPrice(chainId, tokenIn),
+      this.getPrice(chainId, tokenOut),
+    ]);
+
+    return BigInt(
+      Math.ceil(
+        ((((Number(amountOut) * priceOut) / priceIn) * 10 ** decimalsIn) /
+          10 ** decimalsOut) *
+          1.005
+      )
+    );
+  }
+
+  private async getPathId(
+    chainId: string,
+    tokenIn: Address,
+    tokenOut: Address,
+    amountIn?: bigint,
+    amountOut?: bigint
+  ): Promise<string> {
+    if (!amountIn && !amountOut) {
+      throw new Error("amountIn or amountOut must be defined");
+    }
+
+    const swapAdapterConfig = getSwapAdapterConfig(chainId);
+    let estimatedAmountIn =
+      amountIn ??
+      (await this.estimateAmountIn(chainId, tokenIn, tokenOut, amountOut!));
+    const responseQuote = await axios.post(
+      "https://api.odos.xyz/sor/quote/v2",
+      {
+        chainId: swapAdapterConfig.chainIdNum,
+        inputTokens: [
+          {
+            amount: estimatedAmountIn.toString(),
+            tokenAddress: tokenIn,
+          },
+        ],
+        outputTokens: [
+          {
+            proportion: 1,
+            tokenAddress: tokenOut,
+          },
+        ],
+        slippageLimitPercent: 0.5,
+        userAddr: swapAdapterConfig.multicallAdapter,
+      }
+    );
+
+    if (amountOut) {
+      if (responseQuote.data.outTokens[0] < amountOut) {
+        throw new Error("Not enough output tokens");
+      }
+    }
+
+    return responseQuote.data.pathId as string;
+  }
+
+  async constructSwaps(
+    chainId: string,
+    tokenIn: Address,
+    tokenOut: Address,
+    amountIn?: bigint,
+    amountOut?: bigint
+  ): Promise<ExtendedTargetCall[]> {
+    const swapAdapterConfig = getSwapAdapterConfig(chainId);
+    const pathId = await this.getPathId(
+      chainId,
+      tokenIn,
+      tokenOut,
+      amountIn,
+      amountOut
+    );
+
+    const responseTx = await axios.post("https://api.odos.xyz/sor/assemble", {
+      pathId: pathId,
+      simulate: false,
+      userAddr: swapAdapterConfig.multicallAdapter,
+    });
+
+    const targetCalldata: Hex = responseTx.data.transaction.data;
+    const targetContract: Address = responseTx.data.transaction.to;
+    const targetCallValue: bigint = BigInt(responseTx.data.transaction.value);
+    const tokensToSend: TokenToSend[] = responseTx.data.inputTokens.map(
+      (inputToken: { tokenAddress: Address; amount: string }) => ({
+        tokenAmount: {
+          token: inputToken.tokenAddress as Address,
+          amount: inputToken.amount,
+        },
+        destination: responseTx.data.transaction.to,
+      })
+    );
+    const tokensToReceive: TokenAmount[] = responseTx.data.outputTokens.map(
+      (outputToken: { tokenAddress: Address; amount: string }) => ({
+        token: outputToken.tokenAddress as Address,
+        amount: amountOut
+          ? outputToken.amount
+          : Math.floor(Number(outputToken.amount) * 0.995).toString(),
```

Comment:
If this number is related to the `1.005` constant defined above, let's derive them from the same constant